### ISSUE TITLE
[Important] can not update service publish port(host mode) in docker 1.13.0-rc5

### ISSUE THREAD

**Issue by yank1** [docker/docker#30199]
**Description**

**Steps to reproduce the issue:**
1. docker service create --publish target=80,published=12348,protocol=tcp,mode=host nginx test
2. docker service update --publish-rm target=80,published=12348,protocol=tcp,mode=host --publish-add target=80,published=12349,protocol=tcp,mode=host test
3. docker service inspect test

**Describe the results you received:**

```
[
    {
        "ID": "r8z1z994uqtnrkl0x1rsdstaz",
        "Version": {
            "Index": 8456
        },
        "CreatedAt": "2017-01-17T07:53:04.185574076Z",
        "UpdatedAt": "2017-01-17T07:54:41.310011671Z",
        "Spec": {
            "Name": "thirsty_bell",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "daocloud.io/nginx:latest@sha256:b79693b5fdce190faec138b8a1f7053286aaab0d6f521ffa5dca4764c32c596b",
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "MaxFailureRatio": 0
            },
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 12349,
                        "PublishMode": "host"
                    }
                ]
            }
        },
        "PreviousSpec": {
            "Name": "thirsty_bell",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "daocloud.io/nginx:latest@sha256:b79693b5fdce190faec138b8a1f7053286aaab0d6f521ffa5dca4764c32c596b",
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "MaxFailureRatio": 0
            },
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 12348,
                        "PublishMode": "host"
                    }
                ]
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 12348,
                        "PublishMode": "host"
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 80,
                    "PublishedPort": 12348,
                    "PublishMode": "host"
                }
            ]
        },
        "UpdateStatus": {
            "State": "updating",
            "StartedAt": "2017-01-17T07:54:41.309996141Z",
            "CompletedAt": "1970-01-01T00:00:00Z",
            "Message": "update in progress"
        }
    }
]
```

**Describe the results you expected:**

```
"Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 12349,
                        "PublishMode": "host"
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 80,
                    "PublishedPort": 12349,
                    "PublishMode": "host"
                }
            ]
        },
```

**Additional information you deem important (e.g. issue happens only occasionally):**

Can not change publish port in 1.13.0-rc5.
The --publish-rm and --publish-add is not honest.

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc5
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 00:43:46 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc5
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 00:43:46 2017
 OS/Arch:      linux/amd64
 Experimental: false
 ```

**Output of `docker info`:**

```
Containers: 35
 Running: 12
 Paused: 0
 Stopped: 23
Images: 19
Server Version: 1.13.0-rc5
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local rexray
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: i99dhbz7k0f6nmfppjs3sdd0w
 Is Manager: true
 ClusterID: xe3bdcd8ncxz2yt2871yf8n09
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.100.201
 Manager Addresses:
  192.168.100.201:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.703 GiB
Name: yk-centos72-1
ID: U7W3:WJPT:KDOH:J3CY:5NSF:UNEC:MOL7:OTR3:EM3V:UEN5:OAVR:QO4R
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: false
Insecure Registries:
 0.0.0.0/0
 127.0.0.0/8
Registry Mirrors:
 http://dce.m.daocloud.io
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by vdemeester** [comment]
hi @yank1, 

`docker service create --publish target=80,published=12348,protocol=tcp,mode=host nginx test` is wrong, should be `docker service create --name test --publish target=80,published=12348,protocol=tcp,mode=host nginx`, I think, right ?

In your inspect, the Ports is also right:

```
        "Spec": {
            "Name": "thirsty_bell",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "daocloud.io/nginx:latest@sha256:b79693b5fdce190faec138b8a1f7053286aaab0d6f521ffa5dca4764c32c596b",
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "any",
                    "MaxAttempts": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "MaxFailureRatio": 0
            },
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 12349,
                        "PublishMode": "host"
                    }
                ]
            }
        },
``` 

The EndpointSpec has the right Ports, but, for some reason, Endpoint is not updated.. It's definitely reproductible.

/ping @mavenugo @aboch @aaronlehmann @vieux

**Comment by thaJeztah** [comment]
It looks like the client is correctly sending an update request, but SwarmKit keeps trying to update the container; this is logged every 30 seconds;

<details>

```
DEBU[0588] update of service jaoca8r7jhcair3pcs76bysik complete
DEBU[0588] assigning to node vx1pb2g60slhyndawj5ewi8bs   task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] (*worker).Update                              len(assignments)=2 module="node/agent"
DEBU[0588] (*worker).reconcileSecrets                    len(removedSecrets)=0 len(updatedSecrets)=0 module="node/agent"
DEBU[0588] (*worker).reconcileTaskState                  len(removedTasks)=0 len(updatedTasks)=2 module="node/agent"
DEBU[0588] assigned                                      module="node/agent" task.desiredstate=SHUTDOWN task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0588] assigned                                      module="node/agent" task.desiredstate=READY task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] state changed                                 module="node/agent" state.desired=READY state.transition="ASSIGNED->ACCEPTED" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] state changed                                 module="node/agent/taskmanager" state.desired=READY state.transition="ASSIGNED->ACCEPTED" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] task status reported                          module="node/agent"
DEBU[0588] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] task status reported                          module="node/agent"
DEBU[0588] state changed                                 module="node/agent/taskmanager" state.desired=READY state.transition="ACCEPTED->PREPARING" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0588] task status reported                          module="node/agent"
DEBU[0588] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="ASSIGNED->ACCEPTED" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0588] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="RUNNING->RUNNING" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0589] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] task status reported                          module="node/agent"
WARN[0589] Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/4250e40acd71ead65a42f44411ca9ae8a13548b942cdc690740734c44dcde92a-init: exec: "auplink": executable file not found in $PATH
DEBU[0589] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="ACCEPTED->PREPARING" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] (*worker).Update                              len(assignments)=1 module="node/agent"
DEBU[0589] (*worker).reconcileSecrets                    len(removedSecrets)=0 len(updatedSecrets)=0 module="node/agent"
DEBU[0589] (*worker).reconcileTaskState                  len(removedTasks)=1 len(updatedTasks)=0 module="node/agent"
ERRO[0589] failed to deactivate service binding for container test.1.jne7so958l2vgurqcy9q0mjrf  error="network sandbox does not exist for container test.1.jne7so958l2vgurqcy9q0mjrf" module="node/agent"
DEBU[0589] shutdown failed on removal                    error="network sandbox does not exist for container test.1.jne7so958l2vgurqcy9q0mjrf" module="node/agent"
DEBU[0589] container mounted via layerStore: /var/lib/docker/aufs/mnt/4250e40acd71ead65a42f44411ca9ae8a13548b942cdc690740734c44dcde92a
WARN[0589] Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/4250e40acd71ead65a42f44411ca9ae8a13548b942cdc690740734c44dcde92a: exec: "auplink": executable file not found in $PATH
DEBU[0589] state changed                                 module="node/agent/taskmanager" state.desired=READY state.transition="PREPARING->READY" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] task status reported                          module="node/agent"
DEBU[0589] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="PREPARING->READY" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0589] task status reported                          module="node/agent"
DEBU[0589] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="READY->READY" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0590] Sending kill signal 15 to container 330a063a0cd874731810e0c29e78a92fcad7d83fc413087529529bb71efdc747
DEBU[0590] containerd: process exited                    id=330a063a0cd874731810e0c29e78a92fcad7d83fc413087529529bb71efdc747 pid=init status=0 systemPid=1281
ERRO[0590] containerd: deleting container                error=exit status 1: "container 330a063a0cd874731810e0c29e78a92fcad7d83fc413087529529bb71efdc747 does not exist\none or more of the container deletions failed\n"
DEBU[0590] libcontainerd: received containerd event: &types.Event{Type:"exit", Id:"330a063a0cd874731810e0c29e78a92fcad7d83fc413087529529bb71efdc747", Status:0x0, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc42122f370)}
DEBU[0590] Revoking external connectivity on endpoint test.1.qhaj1t7x4fayh3tmx6o29esnl (0e309cd4426b3bf61b64151a726c101fb898f02aa36587aa4b1af2d1ee89b1f3)
DEBU[0590] Releasing addresses for endpoint test.1.qhaj1t7x4fayh3tmx6o29esnl's interface on network bridge
DEBU[0590] ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2)
WARN[0590] Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/889c7c59c32b7272f051caa51195a0b15f021329c43b2e3e90b9a8e5c23ab57e: exec: "auplink": executable file not found in $PATH
DEBU[0590] state changed                                 module="node/agent/taskmanager" state.desired=SHUTDOWN state.transition="RUNNING->SHUTDOWN" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0590] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0590] task status reported                          module="node/agent"
DEBU[0590] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0590] task status reported                          module="node/agent"
DEBU[0590] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="RUNNING->SHUTDOWN" task.id=qhaj1t7x4fayh3tmx6o29esnl
DEBU[0591] (*worker).Update                              len(assignments)=1 module="node/agent"
DEBU[0591] (*worker).reconcileSecrets                    len(removedSecrets)=0 len(updatedSecrets)=0 module="node/agent"
DEBU[0591] (*worker).reconcileTaskState                  len(removedTasks)=0 len(updatedTasks)=1 module="node/agent"
DEBU[0591] assigned                                      module="node/agent" task.desiredstate=RUNNING task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0591] state changed                                 module="node/agent/taskmanager" state.desired=RUNNING state.transition="READY->STARTING" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0591] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0591] task status reported                          module="node/agent"
DEBU[0591] container mounted via layerStore: /var/lib/docker/aufs/mnt/4250e40acd71ead65a42f44411ca9ae8a13548b942cdc690740734c44dcde92a
DEBU[0591] Assigning addresses for endpoint test.1.p6r9dl97jvsbd8567i8q9w76x's interface on network bridge
DEBU[0591] RequestAddress(LocalDefault/172.18.0.0/16, <nil>, map[])
DEBU[0591] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="READY->STARTING" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0591] Assigning addresses for endpoint test.1.p6r9dl97jvsbd8567i8q9w76x's interface on network bridge
DEBU[0591] Programming external connectivity on endpoint test.1.p6r9dl97jvsbd8567i8q9w76x (9787e378ad99bb5de623e1f76ec06253d0209969420048628d673f4a6c40ab6c)
DEBU[0592] sandbox set key processing took 154.060799ms for container 1b18fe1add77d05cb076e9a13b256d0b018068bd39904348257dcebf01c30801
DEBU[0592] libcontainerd: received containerd event: &types.Event{Type:"start-container", Id:"1b18fe1add77d05cb076e9a13b256d0b018068bd39904348257dcebf01c30801", Status:0x0, Pid:"", Timestamp:(*timestamp.Timestamp)(0xc4216003c0)}
DEBU[0592] libcontainerd: event unhandled: type:"start-container" id:"1b18fe1add77d05cb076e9a13b256d0b018068bd39904348257dcebf01c30801" timestamp:<seconds:1484652760 nanos:12956113 >
DEBU[0592] state changed                                 module="node/agent/taskmanager" state.desired=RUNNING state.transition="STARTING->RUNNING" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0592] (*Agent).UpdateTaskStatus                     module="node/agent" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0592] task status reported                          module="node/agent"
DEBU[0592] waiting on events                             module="node/agent/taskmanager" task.id=p6r9dl97jvsbd8567i8q9w76x
DEBU[0592] task status updated                           method="(*Dispatcher).processUpdates" module=dispatcher state.transition="STARTING->RUNNING" task.id=p6r9dl97jvsbd8567i8q9w76x
```

</details>

**Comment by yongtang** [comment]
I think the issue is that `isPortsAllocated()` does not capture host mode:
https://github.com/docker/docker/blob/v1.13.0-rc7/vendor/github.com/docker/swarmkit/manager/allocator/networkallocator/portallocator.go

That caused the EventUpdateService to skip the update of the service stored. Will take a look in SwarmKit.

**Comment by thaJeztah** [comment]
Reopening for now, because https://github.com/docker/docker/issues/30548 fixes this for 1.14, but we need a cherry-pick for 1.13.1

**Comment by vieux** [comment]
closing, it's fixed in both master and 1.13.x
